event.go

```
@@ -0,0 +1,178 @@
+// output data of event or return value of Map if all Filter funcs returned as true.
+func (e *EventEmitter) Execute(serviceID, task string) (*Listener, error) {
+	e.taskServiceID = serviceID
+	e.taskKey = task
+	listener := newListener(e.app, e.gracefulWait)
+	if err := e.app.startServices(e.eventServiceID, serviceID); err != nil {
+		return nil, err
+	}
+	cancel, err := e.listen(listener)
+	if err != nil {
+		return nil, err
+	}
+	listener.cancel = cancel
+	e.app.addListener(listener)
+	return listener, nil
+}
+
+// listen starts listening for events.
+func (e *EventEmitter) listen(listener *Listener) (context.CancelFunc, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	resp, err := e.app.client.ListenEvent(ctx, &core.ListenEventRequest{
+		ServiceID:   e.eventServiceID,
+		EventFilter: e.eventKey,
+	})
+	if err != nil {
+		return cancel, err
+	}
+	go e.readStream(listener, resp)
+	return cancel, nil
+}
+
+// readStream reads listen result stream.
+func (e *EventEmitter) readStream(listener *Listener, resp core.Core_ListenEventClient) {
+	for {
+		e.gracefulWait.Add(1)
+		data, err := resp.Recv()
+		if err != nil {
+			e.gracefulWait.Done()
+			listener.sendError(err)
+			return
+		}
+		event := &Event{
+			Key:  data.EventKey,
+			data: data.EventData,
+		}
+		go e.execute(listener, event)
+	}
+}
+
+// execute executes the task with data returned from Map if all filters
+// are met.
+func (e *EventEmitter) execute(listener *Listener, event *Event) {
+	defer e.gracefulWait.Done()
+
+	for _, filterFunc := range e.filterFuncs {
+		if !filterFunc(event) {
+			return
+		}
+	}
+
+	var data Data
+	if e.mapFunc != nil {
+		data = e.mapFunc(event)
+	} else if err := event.Data(&data); err != nil {
+		e.app.log.Println(errDecodingEventData{err})
+		return
+	}
+
+	if _, err := e.app.execute(e.taskServiceID, e.taskKey, data); err != nil {
+		e.app.log.Println(executionError{e.taskKey, err})
```

Why not return this error to the `Listener.Err` chan?